cavern/src/main/java/org/opencadc/cavern/files/CavernURLGenerator.java

```diff
@@ -148,20 +149,31 @@ public List<Protocol> getEndpoints(VOSURI target, Transfer transfer, Job job, Li
         }
         List<Protocol> ret = null;
         try {
-            Direction dir = transfer.getDirection();
-            PathResolver ps = new PathResolver(nodePersistence, authorizer, true);
-            Node n = ps.getNode(target.getParentURI().getPath());
-            // assume not null and Container already checked by caller (TransferRunner)
-            ContainerNode parent = (ContainerNode) n;
-            Node node = nodePersistence.get(parent, target.getName());
+            PathResolver pr = new PathResolver(nodePersistence, authorizer, true);
+            Node node = pr.getNode(target.getPath());
+            ContainerNode parent;
+            String name;
+            if (node == null) {
+                // maybe a DataNode that needs to be created later. Check the parent (Note: a bit repetitive to check
+                // the entire path again)
+                Node tmp = pr.getNode(target.getParent());
```

Comment:
I don't really like this approach of doing the path twice. Isn't it sufficient to leave it as before and maybe just check before casting? 

Conceptually,  I think it is usually the case that the data node (file) should not exist (the create side effect farther upstream in transfer negotiation is kind of a bug I'd like to fix) unless there was an explicit create DataNode (or a file is being replaced). When I fix that, this code will be quite sub-optimal.

Reply:
Re-worked to avoid redundant calls. I was thinking to to it this way but didn't realize `validateTargetURI` was available and I didn't want to do it just for this.